MAINT: Make diff iterative instead of recursive

madphysicist opened the pull request:
Besides making the code shorter, making diff iterative removes the re-computation of the n-dimensional slices with every recursive call. This change is especially targeted for large values of `n`. Not only does it avoid the re-computation of the indices and stack frames necessary for recursion, but by reassigning the output reference over and over, it allows the intermediate arrays to be delete immediately instead of waiting for the stack to unwind.

This PR also wraps the docs to 72 chars and adds a new regression test.

Since this does not introduce any significant external changes, I will not email the thread.

madphysicist committed:
MAINT: Changed diff to use iterative instead of recursive approach

TST: Added tests for `n` parameter
  Added test for `datetime64` type change
  Added tests for axis normalization
  Added test for subtype handling

DOC: Minor updates to docs:
  Added explanation for `n==0`
  Added documentation describing `datetime64` handling
  Updated formatting
  Added call to normalize_axis_index

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1897,24 +1900,26 @@ def diff(a, n=1, axis=-1):
+    slice1 = tuple(slice(1, None) if i == axis else keep for i in rng)
+    slice2 = tuple(slice(None, -1) if i == axis else keep for i in rng)
+
+    op = operator.ne if a.dtype == np.bool_ else operator.sub
```
Why not `np.not_equal` and `np.subtract`?

madphysicist replied:
Fixed in #9481, forgot to fix here, but did so now.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1897,24 +1900,26 @@ def diff(a, n=1, axis=-1):
+    if n >= a.shape[axis]:
+        shape = list(a.shape)
+        shape[axis] = 0
+        return empty(shape, dtype=a.dtype)
```
This isn't subclass safe

eric-wieser replied:
(whereas previously, it was)

madphysicist replied:
Of course. Fixed. Added a test.

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -658,6 +658,17 @@ def test_nd(self):
+        assert_raises(ValueError, diff, x, n=-1)
+        output = [diff(x, n=n) for n in range(1, 5)]
+        expected = [[1, 1], [0], [], []]
+        assert_(diff(x, n=0) is x)
```
Seems to me that this should maybe return a copy for consistency, but that's not relevant to this pr

madphysicist replied:
The previous behavior is to return the object itself. I documented that and added a test, but I want to avoid making an API change. Instead of a copy, I would probably go with at least an application of `asanyarray`.

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -670,6 +671,20 @@ def test_n(self):
+        Check that subclass is preserved, even across optimization for
+        large `n`.
+        """
+        x = ma.array([[1, 2], [3, 4], [5, 6]])
```
This should actually set a mask too

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1897,24 +1899,25 @@ def diff(a, n=1, axis=-1):
-        da = a[slice1] - a[slice2]
+    if n >= a.shape[axis]:
+        slice1 = tuple(slice(0, 0) if i == axis else keep for i in rng)
+        return a[slice1]
```
I don't think we need this optimization, and I can see it being harmful for a datetime64 array of 1 element, which should return a timedelta

madphysicist replied:
I will add a test to make sure that the time-delta case works out as expected. If it lets me, I may retain the optimization, but with `>` instead of `>=`.

eric-wieser replied:
That won't help. You'll just have the same problem with larger n

madphysicist replied:
The array won't eventually devolve to size zero in one dimension?

eric-wieser replied:
This branch in only runs at the start

madphysicist replied:
Oh, I see. The problem is that the dtype will be wrong, not that the size will be. I'll fix that.

eric-wieser replied:
Correct. Please just remove the optimization, because you can't predict how every subclass and dtype ever will behave.

madphysicist replied:
Will do of course. Same for #9481

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1897,24 +1899,25 @@ def diff(a, n=1, axis=-1):
-    slice1 = tuple(slice1)
-    slice2 = tuple(slice2)
+    rng = range(nd)
+    axis = rng[axis]
```
`axis = normalize_axis_index(axis, nd)`

madphysicist replied:
Thank you for that. I was looking for something like that here: https://stackoverflow.com/q/45364139/2988730. If you post an answer, I will definitely accept it in preference to the `range`-based solution.

eric-wieser replied:
It's also possible your implementation is faster than our current one - so feel free to replace it in another PR, if you profile it.

madphysicist replied:
I took a look at the [implementation](https://github.com/numpy/numpy/blob/260f18c910175bcb6162feb363112929653276aa/numpy/core/src/multiarray/common.h#L142), and I seriously doubt that creating and indexing a `range` has any chance of being faster.

eric-wieser commented:
If it's any consolation, this was the solution I had in mind too until I
thought about datetime64s

On Fri, Jul 28, 2017, 4:15 PM Joseph Fox-Rabinovitz <
notifications@github.com> wrote:

> *@madphysicist* commented on this pull request.
> ------------------------------
>
> In numpy/lib/function_base.py
> <https://github.com/numpy/numpy/pull/9482#discussion_r130114974>:
>
> >
> -    if a.dtype == np.bool_:
> -        da = a[slice1] != a[slice2]
> -    else:
> -        da = a[slice1] - a[slice2]
> +    if n >= a.shape[axis]:
> +        slice1 = tuple(slice(0, 0) if i == axis else keep for i in rng)
> +        return a[slice1]
>
> Will do of course. Same for #9481
> <https://github.com/numpy/numpy/pull/9481>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9482#discussion_r130114974>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ9LKtdOEo9lkzHV5PSy9gyBc7W2JJMks5sSfr_gaJpZM4OmDSa>
> .
>

madphysicist commented:
What consoles me the most is that I don't expect that `n != 1` is ever used much at all to begin with.

madphysicist commented:
@eric-wieser I've removed the optimization, switched to using ` normalize_axis_index`, added a mask to the subtype test and added new tests to show that the axis normalization and `datetime` are handled correctly. I also added a note to the docs about `datetime64->timedelta64`.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1859,23 +1861,29 @@ def diff(a, n=1, axis=-1):
     >>> np.diff(i16_arr)
     array([-1], dtype=int16)
 
+    Some types, like ``datetime64`` will not be preserved at all:
```
This note should go up by the return value docstring, which falsely claims the return is the same dtype

madphysicist replied:
Will do.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1897,24 +1905,21 @@ def diff(a, n=1, axis=-1):
-    else:
-        da = a[slice1] - a[slice2]
+    slice1 = tuple(slice(1, None) if i == axis else keep for i in rng)
+    slice2 = tuple(slice(None, -1) if i == axis else keep for i in rng)
```
Not convinced this is more readable... `keep` is definitely unhelpful here

madphysicist replied:
It also appears to be slower than the original: https://stackoverflow.com/questions/45364139/is-there-a-built-in-method-for-checking-individual-list-indices-in-python/45377475#comment77720595_45364139. I guess I'll revert it.

eric-wieser replied:
Keep the `normalize_axis_index`, because that causes `AxisError` to be thrown

madphysicist replied:
Will do.

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -658,6 +673,49 @@ def test_nd(self):
+            assert_equal(len(out), max(0, len(x) - n))
+
+    def test_times(self):
+        x = np.arange('1066-10-13', 3, dtype=np.datetime64)
```
This line makes no sense to me at all (I know it works). Where is this behaviour documented?

madphysicist replied:
I have no idea either to be honest. I am very bad at working with the time dtypes, to I just went ahead and played with it until it gave me a 3-element array centered on the day of the battle of Hastings.

madphysicist replied:
What is the proper way to do something like this?

mhvk replied:
From my comment on the docstring:
```
np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
```
This seems a little bit more like the normal use of `arange`.

madphysicist commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -6,6 +6,8 @@
 
 import numpy as np
+from numpy import ma
+from numpy.core._internal import AxisError
```
You are right about that. Will fix.

madphysicist commented:
@eric-wieser I believe that I have fixed all the additional comments besides figuring out how the `datetime64` range works.

eric-wieser commented:
> besides figuring out how the datetime64 range works.

Yeah, this probably belongs as a separate issue anyway

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -658,6 +672,49 @@ def test_nd(self):
+    def test_subclass(self):
+        """
+        Check that subclass is preserved, even across optimization for
+        large `n`.
```
No such optimization exists, so perhaps not something to mention

madphysicist replied:
Right, old relic

eric-wieser reviewed (approved):
Apart from a now-outdated comment, all looks good to me

madphysicist commented:
My concern with the range feature is that it is something that is part of the public API, not just a hack I accidentally uncovered. If you are OK keeping it the way it is, I will do so. If not, could you please show me how to fix it?

madphysicist commented:
I've removed the comment. Thanks for taking the time to review this code so thoroughly. Would you mind taking a look at #9481? It is very similar to what I did here, and I have been including most of the changes that we made here as I went along.

madphysicist commented:
I am pretty sure that the test failure is not because of anything I did.

seberg commented:
My guess is, travis updated to python 2.7.13, the release notes mention a change that looks related.

eric-wieser commented:
Can you squash this into a single `MAINT` commit?

madphysicist commented:
@eric-wieser Done.

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1891,30 +1896,33 @@ def diff(a, n=1, axis=-1):
     >>> np.diff(x, axis=0)
     array([[-1,  2,  0, -2]])
 
+    >>> x = np.arange('1066-10-13', 3, dtype=np.datetime64)
```
This does look rather weird. Even though it works, I would suggest to replace it with something that would more directly make sense to a user:
```
np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
```

mhvk reviewed (approved):
This is very nice. Only one small comment about the new `datetime64` example in the docstring - and if that is done, I would suggest making the same change in the te~~x~~st.

madphysicist commented:
@mhvk Agreed and done.

eric-wieser commented:
Thanks @madphysicist, in it goes!
